tensorflow/core/kernels/dynamic_partition_op_gpu.cu.cc

```diff
@@ -0,0 +1,376 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// The algorithm for dynamic partition has the following steps:
+// 1. Let N be the size of partitions. We initialize a new vector indices_in
+//    with the values 0, 1, 2, ..., N-1.
+// 2. We apply cub::DeviceRadixSort::SortPairs to the key - value pairs given
+//    by partitions and indices_in. This will result in two new vectors
+//    partitions_out and indices_out, with partitions_out sorted.
+// 3. The first dimension of outputs[i] is equal to the length of the interval
+//    of i-values in partitions_out. We determine it in two steps:
+//    - compute the starting and ending point of each interval,
+//    - subtract the starting and ending points to find the length.
+//    The result is placed in partition_count.
```

Comment:
Just looking at this again and realized there is a faster / less custom code way of determining the partition counts:  use reduce_by_key from cub: http://nvlabs.github.io/cub/structcub_1_1_device_reduce.html#a303ae673ac32825f95912b4bfff8bef1

The sorted partitions would be the keys and a constant_iterator of 1 would be the values.  It should do less passes over the memory than the current implementation.  Not critical though, so do it if you find it interesting.

Reply:
You are right, that's a better way of doing things. I will create a separate PR for this.

Reply:
Actually, I think it's not that easy. The problem is that there could be wrong indices in partitions, which are larger than num_partitions. cub::ReduceByKey does not check the input for correctness and assumes the output array d_unique_out is large enough to contain all the distinct keys. But this need not be the case if the input is wrong. This would result in a seg fault or some other memory corruption. On the other hand, my custom kernel simply ignores out-of-bounds indices.
 It would be possible to do a check first, but then I would have to wait for the check to complete, bring the result to the CPU, and decide whether I should continue the computation. I think it's easier to keep the version which just ignores the out-of-bound indices.

Reply:
Maybe you could use custom iterators (instead of pointers).  The output iterators could do a check for the position being in bounds and only then doing the write.  This wouldn't involve any extra kernel launches or host communication.

Reply:
You can find an example of a custom output iterator here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/transform_output_iterator.h